sbin/health-checker.in

```diff
@@ -71,77 +58,77 @@ telem_send_record()
     fi
 }
 
-save_working_snapshot()
-{
-    set_btrfs_id
-
-    if [ ${BTRFS_ID_DEFAULT} -eq ${BTRFS_ID_CURRENT} ]; then
-	echo "LAST_WORKING_BTRFS_ID=${BTRFS_ID_DEFAULT}" > $STATE_FILE
-	echo "LAST_WORKING_SNAPSHOT=${SNAPSHOT_DEFAULT}" >> $STATE_FILE
-    fi
-}
-
-rollback()
-{
-    . ${STATE_FILE}
-    mount -o remount,rw /.snapshots
-    btrfs subvolume set-default ${LAST_WORKING_BTRFS_ID} /.snapshots
-    if [ $? -ne 0 ]; then
-        create_log user.crit "ERROR: btrfs set-default $BTRFS_ID_DEFAULT failed!"
-	telem_send_payload 1
-        exit 1
+systemd-bless-boot() {
+    if [ -x /usr/lib/systemd/systemd-bless-boot ]; then
+        /usr/lib/systemd/systemd-bless-boot "$@"
     fi
 }
 
-stop_services()
+error_decision()
 {
-    # Stop all services
-    for script in ${PLUGINDIR}/* ${USR_LOCAL_PLUGINDIR}/*; do
-        if [ -f ${script} ]; then
-            ${script} stop
+    local status should_reboot entries default_entry selected_entry current_snapshot tries
+    read -r tries < /etc/kernel/tries
+    # systemd-bless-boot returns:
+    #   clean: boot counting is not in effect
+    #   good:  this entry booted fine. Since we are calling this at boot,
+    #          it shouldn't be possible to return "good"
+    #   bad:   this entry was already marked as bad after 3 tries
+    #   indeterminate: when an entry is neither good or bad, i.e.
+    #                  we are still trying to boot 3 times
+    status=$(systemd-bless-boot status)
+    # systemd-boot fill the EFI variables with the info we need
+    
+    # Get the entry depending on the current booted snapshot
+    current_snapshot=$(findmnt --output OPTIONS --noheadings / | sed -e 's|.*subvol=/@/.snapshots/\([0-9]\+\)/snapshot.*|\1|g')
+    # We do not want a file that does not exist
+    shopt -s nullglob
+    selected_entry=(/boot/efi/loader/entries/*[a-z]-"$current_snapshot"+[[:digit:]]-[[:digit:]].conf /boot/efi/loader/entries/*[a-z]-"$current_snapshot".conf)
```

Comment:
This should ideally not hardcode `/boot/efi/loader/entries` or the format of entries.

Would be nice to use `bootctl list --json=pretty | jq -r '.[]|select(.isSelected).id'` here, but I guess that does not work with grub?

Reply:
there is a bli.mod in grub2 (https://www.gnu.org/software/grub/manual/grub/html_node/bli.html) but only populate two EFI variables. If this gets improved we can use bootctl

Reply:
I have a patch for grub2 that populates `LoaderBootCountPath` since it need by health-checker. I'll look up which other variables bootctl uses.

Reply:
This would be the best solution. The variables are described here: https://systemd.io/BOOT_LOADER_INTERFACE/ and I think that .isSelected comes from LoaderEntrySelected. There are others that would help, for example LoaderEntryDefault or LoaderConfigTimeout